File: src/test/java/com/robohorse/robopojogenerator/generator/processors/ClassProcessorTest.java
package com.robohorse.robopojogenerator.generator.processors;

import com.robohorse.robopojogenerator.generator.common.ClassDecorator;
import com.robohorse.robopojogenerator.generator.common.ClassItem;
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import testutils.JsonReader;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import static org.junit.Assert.*;
import static org.mockito.Mockito.when;

/**
 * Created by vadim on 02.10.16.
 */

public class ClassProcessorTest {
    private JsonReader jsonReader = new JsonReader();

    @InjectMocks
    ClassProcessor classProcessor;
    @Mock
    ClassGenerateHelper classGenerateHelper;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void testSingleObjectGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("single_object.json");
        final String name = "Response";
        final Set<ClassItem> classItemSet = new HashSet<>();

        when(classGenerateHelper.getClassName(name))
                .thenReturn(name);

        classProcessor.proceed(jsonObject, name, classItemSet);
        assertTrue(classItemSet.size() == 1);

        Iterator iterator = classItemSet.iterator();
        ClassItem classItem = (ClassItem) iterator.next();
        assertEquals(classItem.getClassName(), name);

        final Map<String, ClassDecorator> fields = classItem.getClassFields();
        assertNotNull(fields);

        for (String key : jsonObject.keySet()) {
            assertTrue(fields.containsKey(key));
        }
    }

    @Test
    public void testInnerObjectGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("inner_json_object.json");
        final JSONObject innerJsonObject = jsonObject.optJSONObject("data");
        final String name = "Response";
        final Set<ClassItem> classItemSet = new HashSet<>();

        when(classGenerateHelper.getClassName(name))
                .thenReturn(name);

        classProcessor.proceed(jsonObject, name, classItemSet);
        assertTrue(classItemSet.size() == 2);

        for (ClassItem classItem : classItemSet) {
            final Map<String, ClassDecorator> fields = classItem.getClassFields();
            assertNotNull(fields);

            if (name.equalsIgnoreCase(classItem.getClassName())) {
                for (String key : jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                }
            } else {
                for (String key : innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                }
            }
        }
    }

    @Test
    public void testEmptyArrayGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("empty_array.json");
        final String name = "Response";
        final Set<ClassItem> classItemSet = new HashSet<>();

        when(classGenerateHelper.getClassName(name))
                .thenReturn(name);

        classProcessor.proceed(jsonObject, name, classItemSet);
        assertTrue(classItemSet.size() == 1);

        Iterator iterator = classItemSet.iterator();
        ClassItem classItem = (ClassItem) iterator.next();
        assertEquals(classItem.getClassName(), name);

        final Map<String, ClassDecorator> fields = classItem.getClassFields();
        assertNotNull(fields);

        for (String key : jsonObject.keySet()) {
            assertTrue(fields.containsKey(key));
        }

        final ClassDecorator targetObjectType = classItem.getClassFields().get("data");

        assertEquals("List<Object>", targetObjectType.getJavaItem());
    }

    @Test
    public void testIntegerArrayGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("array_with_primitive.json");
        final String name = "Response";
        final String targetType = "List<Integer>";

        final Set<ClassItem> classItemSet = new HashSet<>();

        when(classGenerateHelper.getClassName(name))
                .thenReturn(name);

        classProcessor.proceed(jsonObject, name, classItemSet);
        assertTrue(classItemSet.size() == 1);

        Iterator iterator = classItemSet.iterator();
        ClassItem classItem = (ClassItem) iterator.next();
        assertEquals(classItem.getClassName(), name);

        final Map<String, ClassDecorator> fields = classItem.getClassFields();
        assertNotNull(fields);

        for (String key : jsonObject.keySet()) {
            assertTrue(fields.containsKey(key));
        }

        final ClassDecorator actualType = classItem.getClassFields().get("data");

        assertEquals(targetType, actualType.getJavaItem());
    }

    @Test
    public void testInnerArrayObjectGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("array_with_jsonobject.json");
        final JSONArray innerJsonArray = jsonObject.optJSONArray("data");
        final JSONObject innerJsonObject = innerJsonArray.getJSONObject(0);
        final String name = "Response";
        final String targetType = "List<DataItem>";

        final Set<ClassItem> classItemSet = new HashSet<>();

        when(classGenerateHelper.getClassName(name))
                .thenReturn(name);
        when(classGenerateHelper.getClassNameWithItemPostfix(Mockito.anyString()))
                .thenReturn("DataItem");

        classProcessor.proceed(jsonObject, name, classItemSet);
        assertTrue(classItemSet.size() == 2);

        for (ClassItem classItem : classItemSet) {
            final Map<String, ClassDecorator> fields = classItem.getClassFields();
            assertNotNull(fields);

            if (name.equalsIgnoreCase(classItem.getClassName())) {
                for (String key : jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                    final ClassDecorator actualType = classItem.getClassFields().get("data");
                    final String javaItem = actualType.getJavaItem();
                    assertEquals(targetType, javaItem);
                }
            } else {
                for (String key : innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                }
            }
        }
    }
}

File: generator/src/test/java/com/robohorse/robopojogenerator/generator/processing/ClassProcessorTest.kt
package com.robohorse.robopojogenerator.generator.processing

import com.robohorse.robopojogenerator.generator.consts.common.ClassItem
import com.robohorse.robopojogenerator.generator.consts.common.JsonModel.JsonItem
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper
import com.robohorse.robopojogenerator.utils.JsonReader
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.RelaxedMockK
import org.json.JSONObject
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

internal class ClassProcessorTest {
    private val jsonReader: JsonReader = JsonReader()

    @InjectMockKs
    lateinit var classProcessor: ClassProcessor

    @RelaxedMockK
    lateinit var classGenerateHelper: ClassGenerateHelper

    @BeforeTest
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun testSingleObjectGeneration_isCorrect() {
        val jsonObject = jsonReader.read("single_object.json") as JSONObject
        val name = "Response"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        val classItemMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val iterator = classItemMap.values.iterator()
        val (className, _, _, _, fields) = iterator.next()
        assertEquals(className, name)
        assertNotNull(fields)
        for (key in jsonObject.keySet()) {
            assertTrue(fields.containsKey(key))
        }
    }

    @Test
    fun testInnerObjectGeneration_isCorrect() {
        val jsonObject = jsonReader.read("inner_json_object.json") as JSONObject
        val innerJsonObject = jsonObject.optJSONObject("data")
        val name = "Response"
        every { classGenerateHelper.formatClassName(name) }.returns(name)

        val classItemMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 2)
        for ((className, _, _, _, fields) in classItemMap.values) {
            assertNotNull(fields)
            if (name.equals(className, ignoreCase = true)) {
                for (key in jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                }
            } else {
                for (key in innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                }
            }
        }
    }

    @Test
    fun testEmptyArrayGeneration_isCorrect() {
        val jsonObject = jsonReader.read("empty_array.json") as JSONObject
        val name = "Response"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        val classItemMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val iterator: Iterator<*> = classItemMap.values.iterator()
        val (className, _, _, _, fields) = iterator.next() as ClassItem
        assertEquals(className, name)
        assertNotNull(fields)
        for (key in jsonObject.keySet()) {
            assertTrue(fields.containsKey(key))
        }
        val targetObjectType = fields["data"]
        assertEquals("List<Object>", targetObjectType!!.getJavaItem())
    }

    @Test
    fun testIntegerArrayGeneration_isCorrect() {
        val jsonObject = jsonReader.read("array_with_primitive.json") as JSONObject
        val name = "Response"
        val targetType = "List<Integer>"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        val classItemMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val iterator: Iterator<*> = classItemMap.values.iterator()
        val (className, _, _, _, fields) = iterator.next() as ClassItem
        assertEquals(className, name)
        assertNotNull(fields)
        for (key in jsonObject.keySet()) {
            assertTrue(fields.containsKey(key))
        }
        val actualType = fields["data"]
        assertEquals(targetType, actualType!!.getJavaItem())
    }

    @Test
    fun testInnerArrayObjectGeneration_isCorrect() {
        val jsonObject = jsonReader.read("array_with_jsonobject.json") as JSONObject
        val innerJsonArray = jsonObject.optJSONArray("data")
        val innerJsonObject = innerJsonArray.getJSONObject(0)
        val name = "Response"
        val targetType = "List<DataItem>"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        every { classGenerateHelper.getClassNameWithItemPostfix(any()) }.returns("DataItem")
        val classItemMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 2)
        for ((className, _, _, _, fields) in classItemMap.values) {
            assertNotNull(fields)
            if (name.equals(className, ignoreCase = true)) {
                for (key in jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                    val actualType = fields["data"]
                    val javaItem = actualType!!.getJavaItem()
                    assertEquals(targetType, javaItem)
                }
            } else {
                for (key in innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                }
            }
        }
    }

    @Test
    fun testNestedArrayElementsConsolidateObjectProperties() {
        val jsonObject = jsonReader.read("nested_array_elements.json") as JSONObject
        val name = "Response"
        val classItemMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val classFields = classItemMap.values.firstOrNull()?.classFields
        assertNotNull(classFields)
        // when using "itemMap.putAll(innerItemsMap);" element "C" will be missing as newer instances of objects in
        // innerItemsMap of the same name/type overwrite existing map (itemMap) entries
        listOf("Outsides", "Outside", "Insides", "Inside", "A", "B", "C").listIterator().forEach {
            assertTrue(classFields.containsKey(it), "Missing element: $it")
        }
    }
}
